### PR TITLE
mgmt: hawkbit: define attributes in the user application

### LINKED ISSUE #37640: mgmt: hawkbit: allow user defintion of device attributes.
**Is your enhancement proposal related to a problem? Please describe.**
The problem here is that the device attributes are set in the hawkbit code.
It would be better if these attributes are user configurable.

**Describe the solution you'd like**
The hawkbit user should be able to configure the device attributes in their own code.
The usage of default values is fine for a sample application, but is not adequate for production.
Maybe add a Kconfig option like `CONFIG_HAWBIT_USERDEFINED_ATTRIBUTES` and then
require the user to provide an implementation for
```C
struct hawkbit_cfg_data {
	const char *VIN;
	const char *hwRevision;
};
```
and the necessary serialization functions. This would require having a hook that gets called when
hawkbit is inatialized.

**Describe alternatives you've considered**
The only alternative I can see is patching the zephyr source for your own needs.

**Additional context**
Add any other context or graphics (drag-and-drop an image) about the feature request here.

### PR DESCRIPTION
**Pull request by maass-hamburg**
Attributes can now be defined in the user application by the use of a callback function. 
Resolves: #37640 

image

### CODE REVIEW

**Comment by maass-hamburg** [comment]
ping

**Comment by maass-hamburg** [comment]
ping @ycsin

**Comment by maass-hamburg** [comment]
just a rebase to the current main

**Comment by ycsin** [comment]
I haven't really looked at this in details yet but it seems to expose a little too much to the public than I would like, i.e. all the structs and stuff that used to be internal.

I wonder if it is possible to just have a mechanism for hawkBit to get the formatted json of the attributes from the app in the form of a buffer and sent it to the server?

**Comment by maass-hamburg** [comment]
> I haven't really looked at this in details yet but it seems to expose a little too much to the public than I would like, i.e. all the structs and stuff that used to be internal.

they are needed if the user wants to define its own config callback. but maybe they could be put in a seperate header file and not in `include/zephyr/mgmt/hawkbit.h` maybe `include/zephyr/mgmt/hawkbit_config.h` or create a new folder `include/zephyr/mgmt/hawkbit/` which contains `include/zephyr/mgmt/hawkbit/hawkbit.h` and `include/zephyr/mgmt/hawkbit/config.h` ?
 
> I wonder if it is possible to just have a mechanism for hawkBit to get the formatted json of the attributes from the app in the form of a buffer and sent it to the server?

It might be, but then we can't use the json library to encode `struct json_obj_descr json_cfg_descr`, only for `struct json_obj_descr json_cfg_data_descr`. As it is currently not possible to add a already encoded object to a `json_obj_descr`.

**Comment by maass-hamburg** [comment]
@d3zd3z should I put the change of the json lib into a seperate pr?

File: subsys/mgmt/hawkbit/hawkbit.c [hunk 1]
```diff
@@ -827,6 +857,7 @@ static bool send_request(enum http_method method, enum hawkbit_http_request type
 	hb_context.http_req.header_fields = (const char **)headers;
 #endif
 	hb_context.final_data_received = false;
+	hb_context.status_buffer_size = sizeof(hb_context.status_buffer);
```
Comment (ycsin) [inline subsys/mgmt/hawkbit/hawkbit.c:860]: Good catch! This looks like a potential bug fix and should probably be in another PR, so that we can backport it to earlier releases.

I also wonder if it make sense to remove `status_buffer_size` from the struct and simply use `sizeof(hb_context.status_buffer) - 1` instead of `hb_context.status_buffer_size - 1`?

**Review by ycsin — COMMENTED** [review]
> @d3zd3z should I put the change of the json lib into a seperate pr?

It might be a good idea to create a PR for that change and add proper tests/samples for it so that it is easier to review

File: subsys/mgmt/hawkbit/hawkbit.c [hunk 2]
```diff
@@ -840,10 +871,18 @@ static bool send_request(enum http_method method, enum hawkbit_http_request type
 		break;
 
 	case HAWKBIT_CONFIG_DEVICE:
-		memset(&cfg, 0, sizeof(cfg));
+		struct hawkbit_cfg cfg = {0};
```
Comment (ycsin) [inline subsys/mgmt/hawkbit/hawkbit.c:874]: What do you think about sending only the minimal info in default hawkBit implementation (i.e. the existing stuff like `VIN` (I forgot if `hwRevision` is strictly required)), and `merge` the custom attributes in an additional steps?

File: subsys/mgmt/hawkbit/hawkbit.c [hunk 3]
```diff
@@ -840,10 +871,18 @@ static bool send_request(enum http_method method, enum hawkbit_http_request type
...
+
+		char cfg_data[CONFIG_HAWKBIT_ADDITIONAL_STATUS_BUFFER_SIZE] = {0};
+
+		ret = hawkbit_config_device_data_cb_handler(device_id, cfg_data, sizeof(cfg_data));
+		if (ret) {
+			LOG_ERR("Can't encode the JSON script (HAWKBIT_CONFIG_DEVICE, data): %d",
```
Comment (ycsin) [inline subsys/mgmt/hawkbit/hawkbit.c:880]: Please try to adopt the pattern of #69229 if possible

**Review by ycsin — COMMENTED** [review]
probably needs a rebase

File: subsys/mgmt/hawkbit/hawkbit.c (hunk 1 above)
Comment (maass-hamburg) [reply subsys/mgmt/hawkbit/hawkbit.c:860]: this is now addressed in #70355

File: subsys/mgmt/hawkbit/hawkbit.c (hunk 2 above)
Comment (maass-hamburg) [reply subsys/mgmt/hawkbit/hawkbit.c:876]: `VIN` and `hwRevision` are both not strictly required by hawkBit. But I would still like to send `VIN` as our default configuration. `hwRevision` will be removed from the default configuration.

File: subsys/mgmt/hawkbit/hawkbit.c [hunk 4]
```diff
@@ -44,7 +44,7 @@ LOG_MODULE_REGISTER(hawkbit, CONFIG_HAWKBIT_LOG_LEVEL);
 #define RECV_BUFFER_SIZE 640
 #define URL_BUFFER_SIZE 300
 #define SHA256_HASH_SIZE 32
-#define STATUS_BUFFER_SIZE 200
+#define STATUS_BUFFER_SIZE (100 + CONFIG_HAWKBIT_ADDITIONAL_STATUS_BUFFER_SIZE)
```
Comment (m-byte) [inline subsys/mgmt/hawkbit/hawkbit.c:47]: I feel like it would be better to move the whole STATUS_BUFFER_SIZE to Kconfig. Otherwise, there is this weird semi-magic number.
Maybe add a recommendation to the help text so users know that they shouldn't set that value below 100

Comment (maass-hamburg) [reply subsys/mgmt/hawkbit/hawkbit.c]: the 100 was just the half of the default value, it can also be set below 100

**Comment by maass-hamburg** [comment]
@ycsin With #70383 I want to go back to let the user encode `struct json_obj_descr json_cfg_descr` instead of just `struct json_obj_descr json_cfg_data_descr`, because it only contains of `data` and `mode` and no internal stuff has to be exposed. So the user can also set the mode and we don't need the change in the json lib.

**Comment by maass-hamburg** [comment]
just a rebase to resolve conflicts

File: subsys/mgmt/hawkbit/Kconfig [hunk 5]
```diff
@@ -79,6 +79,19 @@ config HAWKBIT_DDI_SECURITY_TOKEN
 	  Authentication security token for the configured Hawkbit DDI
 	  authentication mode.
 
+config HAWKBIT_CUSTOM_ATTRIBUTES
+	bool "Custom device attributes"
+	help
+	  Use custom definition of device attributes.
+
+config HAWKBIT_STATUS_BUFFER_SIZE
+	int "Hawkbit status buffer size"
```
Comment (ycsin) [inline subsys/mgmt/hawkbit/Kconfig:88]:
```suggestion
	int "hawkBit status buffer size"
```

**Review by ycsin — COMMENTED** [review]
The approach in this PR should work, I just wonder if we can keep the default implementation simple (only send `device_id`) but provide more flexibility by having a mechanism (another step or public function, if enabled) to "merge" any custom attributes to the record in the server, i.e. the reason of last reboot (reset cause), die temp, stack watermark, ...

File: samples/subsys/mgmt/hawkbit/prj.conf [hunk 6]
```diff
@@ -58,3 +58,6 @@ CONFIG_LOG_BUFFER_SIZE=4096
 
 #Generate HEX output
 CONFIG_BUILD_OUTPUT_HEX=y
+
+#Use custom attributes for Hawkbit
```
Comment (ycsin) [inline samples/subsys/mgmt/hawkbit/prj.conf:62]: hawkBit

**Comment by maass-hamburg** [comment]
> The approach in this PR should work, I just wonder if we can keep the default implementation simple (only send `device_id`) but provide more flexibility by having a mechanism (another step or public function, if enabled) to "merge" any custom attributes to the record in the server.

As sending the `device_id` as `VIN` is not required by hawkBit. hawkBit doesn't use for anything special. If the user doesn't want to send that, he should be able to do so.

**Comment by maass-hamburg** [comment]
> The approach in this PR should work, I just wonder if we can keep the default implementation simple (only send `device_id`) but provide more flexibility by having a mechanism (another step or public function, if enabled) to "merge" any custom attributes to the record in the server, i.e. the reason of last reboot (reset cause), die temp, stack watermark, ...

This problem is that the device attributes are only send to the server if the server requests it. This normally only happens once. the hawkBit server will only request it again, if you go in the user interface and click there on `Request attributes update`. It could be done to send it without the server requesting but that's outside of the scope of this PR .

File: include/zephyr/mgmt/hawkbit.h [hunk 7]
```diff
@@ -36,6 +36,30 @@ enum hawkbit_response {
...
+						  const size_t buffer_size);
+
+/**
+ * @brief Set the custom data callback.
+ * @details This function is used to set the custom data callback.
+ * The callback is used to provide the custom data to the hawkBit server.
+ * @param cb The callback function.
+ * @retval 0 on success.
+ * @retval -EINVAL if the callback is NULL.
+ */
```
Comment (ycsin) [inline include/zephyr/mgmt/hawkbit.h:60]: We usually have an empty line between `brief`, `details`, `param` and `return`/`retval`, but since `return` is already used in `hawkbit_probe`, it should be used here as well for consistency

**Review by ycsin — COMMENTED** [review]
This patch works - Have you considered doing this kind of one time configurations during `hawkbit_init` by extending it to accept an config struct? The same idea can also applies to #68806

**Comment by maass-hamburg** [comment]
> This patch works - Have you considered doing this kind of one time configurations during `hawkbit_init` by extending it to accept an config struct? The same idea can also applies to #68806

I would prefer to not do it that way, as most subsystems in zephyr are also using this separate approach when it comes to setting callbacks.

**Review by ycsin — DISMISSED** [review]
The doc can use some [improvements](https://github.com/zephyrproject-rtos/zephyr/pull/69293/files#r1561963951), LGTM in general

File: include/zephyr/mgmt/hawkbit.h (hunk 7 above)
Comment (maass-hamburg) [reply include/zephyr/mgmt/hawkbit.h]: is now changed

File: samples/subsys/mgmt/hawkbit/src/main.c [hunk 8]
```diff
@@ -20,6 +21,40 @@
...
+	JSON_OBJ_DESCR_PRIM(struct hawkbit_cfg, mode, JSON_TOK_STRING),
+	JSON_OBJ_DESCR_OBJECT(struct hawkbit_cfg, data, json_cfg_data_descr),
+};
+
+int hawkbit_new_config_data_cb(const char *device_id, uint8_t *buffer, const size_t buffer_size)
+{
+	struct hawkbit_cfg cfg = {
+		.mode = "merge",
+		.data.VIN = device_id,
+		.data.customAttr = "Hello World!",
```
Comment (ycsin) [inline samples/subsys/mgmt/hawkbit/src/main.c:50]: good to attach a screenshot of this in the web UI

**Review by ycsin — DISMISSED** [review]

File: samples/subsys/mgmt/hawkbit/src/main.c (hunk 8 above)
Comment (maass-hamburg) [reply samples/subsys/mgmt/hawkbit/src/main.c:50]: image

**Comment by maass-hamburg** [comment]
rebased to resolve conflicts

**Review by ycsin — APPROVED** [review]
Re-ack

**Review by carlescufi — APPROVED** [review]
LGTM
